### DISCUSSION TITLE
As part of TLS communication, is extended-key-usage and SAN validation enabled?

### DISCUSSION THREAD

**Discussion by mohitvaid** [syslog-ng/syslog-ng#4094] [Q&A] [answered]
Hi Team,

Can you please confirm if as part of TLS communication(peer-verify="yes"), is extended-key-usage and SAN validation enabled?

@MrAnno @tobstarr

**Comment by ramtech123** [comment]
Hi Team,
We are also looking for similar feature where it can support validation of EKU fields and Basic Constraint (CA flag = true in CA certificate).

**Comment by petiepooo** [comment]
It looks like syslog-ng OSE just calls OpenSSL's SSL_set_verify() function with all three of these flags enabled:

> SSL_VERIFY_PEER
The server sends a client certificate request to the client. The certificate returned (if any) is checked. If the verification process fails, the TLS/SSL handshake is immediately terminated with an alert message containing the reason for the verification failure.

> SSL_VERIFY_FAIL_IF_NO_PEER_CERT
If the client did not return a certificate, the TLS/SSL handshake is immediately terminated with a "handshake failure" alert.

> SSL_VERIFY_CLIENT_ONCE
Only request a client certificate on the initial TLS/SSL handshake. Do not ask for a client certificate again in case of a renegotiation.

I don't see that it sets a different verify_callback function, and reading the manpage for the default verify_callback provided by OpenSSL, it just says "At each level signatures and issuer attributes are checked."  Have you tested for the result you want if the right EKU or SAN is not set?

**Comment by ramtech123** [comment, marked as answer]
Hello @petiepooo ,

I have total seven scenarios related to certificate chain validation, out of which five are successful and two fail.

Below is my syslog-ng client configuration
```
destination d_tls_network_1 {
	network(
		"syslog.server.com"
        port(3443)
        transport("tls")
        tls(
            ca-file("/etc/syslog-ng/certs/syslogCACert.pem")
            peer-verify(yes)
        )
	);
};
```
For testing purpose, I'm using openssl s_server as below
```
openssl s_server -accept 3443 -cert server.cert -key server.key
```

Below are the scenarios and test status with different combination of BasicConstraints (of CA), ExtendedKeyUsage (EKU) and CN/SAN attributes of the server certificate

Sl.No | Scenario   (TLS Syslog Server certificate) | Syslog-ng Client Behavior | Test Result | Error/Warning in Syslog-ng logs
-- | -- | -- | -- | --
1 | Invalid issuing certificate - CA flag not set in Basic Constraints | Connection rejected | **PASS** | Invalid certificate found in chain, basicConstraints.ca is unset in non-leaf certificate
2 | Invalid issuing certificate - CA has no Basic Constraints extension| Connection rejected | **PASS** | Invalid certificate found in chain, basicConstraints.ca is unset in non-leaf certificate
3 | Valid certificate path, but leaf certificate has invalid ExtendedKeyUsage (has usages other than `serverAuth`). The extension is not marked as critical. (Refer scenario.8 when extension is critical) | Connection succeeded with warning | **FAIL** (should be able to configure it to enforce EKU purpose, even if it is not critical) | Certificate valid, but purpose is invalid
4 | Valid certificate path, leaf certificate has no ExtendedKeyUsage extension | Connection succeeded | **~FAIL~ PASS** (per clarification received, not to enforce if extension does not exist) | None
5 | Valid certificate chain, syslog-ng configured with hostname matching the CN of server certificate. No DNS and IP SAN attributes present in the certificate. EKU set to `serverAuth` | Connection succeeded | **PASS** | None
6 | Valid certificate chain, syslog-ng configured with hostname matching CN of server certificate. Valid DNS/IP SAN attributes exist in the certificate. EKU set to `serverAuth` | Connection rejected | **PASS**          (As per https://www.rfc-editor.org/rfc/rfc2818#section-3.1, if SAN DNS/IP   is present, CN not to be used) | Certificate subject does not match configured hostname; hostname='cn-syslog.server.com',   certificate='san-syslog.server.com'
7 | Valid certificate chain, syslog-ng configured with hostname matching SAN of server certificate. Valid DNS/IP SAN attributes exist in the certificate. EKU set to `serverAuth`|Connection succeeded | **PASS** | None
8 | Variation of scenario.3, in this case the EKU extension is marked as critical | Connection rejected | **PASS** | Certificate validation failed; unhandled critical extension 

I'm looking for resolution of two failing scenarios.

**Reply by ramtech123** [reply]
Created this feature request - https://github.com/syslog-ng/syslog-ng/issues/4103

**Comment by ramtech123** [comment]
@MrAnno ,
Will you be able to pick up https://github.com/syslog-ng/syslog-ng/issues/4103 for implementation? Unfortunately I'll not be able to contribute as I'm not into C/C++. Thanks.
